feat: adding proxy for upgradeability, slither and solidity-coverage

julianmrodri opened the pull request:
* Adds ProxyAdmin and Proxy factory for Upgradeability
* Add Pausable capabilities
* Adds slither and solidity-coverage dependencies. Fixes slither issues with Reentrancy.

Closes #17  and #16  and #15

ilanolkies commented on migrations/2_deploy_nfts_contract.js:
```diff
@@ -0,0 +1,58 @@
+
+const { encodeCall } = require('@openzeppelin/upgrades');
+
+//const RNS_TESTNET_MULTI_SIG = '0x0';
+//const RNS_MULTI_SIG = '0x0';
```
```suggestion
//const RM_TESTNET_MULTI_SIG = '0x0';
//const RM_MULTI_SIG = '0x0';
```

julianmrodri replied:
Good catch! But I will go with RMKT instead of RM!
RMKT_TESTNET_MULTI_SIG
RMKT_MULTI_SIG

ilanolkies commented on migrations/2_deploy_nfts_contract.js:
```diff
@@ -0,0 +1,58 @@
+      await proxyFactory.deploy(salt, simplePlacementsV1.address, proxyAdmin.address, data);
+
+      /*if (network === 'testnet') {
+        await proxyAdmin.transferOwnership(RNS_TESTNET_MULTI_SIG);
```
```suggestion
        await proxyAdmin.transferOwnership(RM_TESTNET_MULTI_SIG);
```

ilanolkies commented on migrations/2_deploy_nfts_contract.js:
```diff
@@ -0,0 +1,58 @@
+      /*if (network === 'testnet') {
+        await proxyAdmin.transferOwnership(RNS_TESTNET_MULTI_SIG);
+      } else if (network === 'mainnet') {
+        await proxyAdmin.transferOwnership(RNS_MULTI_SIG);
```
```suggestion
        await proxyAdmin.transferOwnership(RM_MULTI_SIG);
```

ilanolkies commented on migrations/2_deploy_nfts_contract.js:
```diff
@@ -0,0 +1,58 @@
+      const owner = await proxyAdmin.owner().then(r => r.toLowerCase());
+
+      /*if (network === 'testnet') {
+        assert.equal(owner, RNS_TESTNET_MULTI_SIG);
```
```suggestion
        assert.equal(owner, RM_TESTNET_MULTI_SIG);
```

ilanolkies commented on migrations/2_deploy_nfts_contract.js:
```diff
@@ -0,0 +1,58 @@
+      /*if (network === 'testnet') {
+        assert.equal(owner, RNS_TESTNET_MULTI_SIG);
+      } else if (network === 'mainnet') {
+        assert.equal(owner, RNS_MULTI_SIG);
```
```suggestion
        assert.equal(owner, RM_MULTI_SIG);
```

ilanolkies commented on contracts/ERC721SimplePlacementsV1.sol:
```diff
@@ -52,11 +56,12 @@ contract ERC721SimplePlacements is Context, ERC677TransferReceiver, IERC777Recip
         ERC1820_REGISTRY.setInterfaceImplementer(address(this), keccak256("ERC777Token"), address(this));
         ERC1820_REGISTRY.setInterfaceImplementer(address(this), keccak256("ERC20Token"), address(this));
         ERC1820_REGISTRY.setInterfaceImplementer(address(this), keccak256("ERC777TokensRecipient"), address(this));
+        Ownable.initialize(sender);
```
Minor detail, I would go for one of this two options
- Use `msg.sender` if the expected is to set the sender as the owner and no parameter
- Use `owner` as parameter name, to clarify the role of that address

julianmrodri replied:
Good point. I think it would be easier to use `msg.sender();`. Then do the transfership of Owner to the Multisig on the Mainnet.

julianmrodri replied:
I will go with `owner`. I need the extra parameter to avoid clashing with the `initialize` of the Ownable contract which also receives an address.I will rename the parameter.

ilanolkies commented on migrations/2_deploy_nfts_contract.js:
```diff
@@ -0,0 +1,58 @@
+        this.tokenAddress = '0xca0a477e19bac7e0e172ccfd2e3c28a7200bdb71';
+      } 
+
+      const proxyFactory = await deployer.deploy(ProxyFactory);
```
RNS team already deployed a Proxy Factory. You can reuse it.

Testnet: [0xbb71c17b28baf2b9e7f5b31660ecf758113b3fef](https://explorer.testnet.rsk.co/address/0xbb71c17b28baf2b9e7f5b31660ecf758113b3fef)
Mainnet: [0x30569c0999f0c7fd7be9957ac4b4dee803cac293](https://explorer.rsk.co/address/0x30569c0999f0c7fd7be9957ac4b4dee803cac293)

julianmrodri replied:
Will reuse. I will harcode it in the Script for testnet and only create it for development.

julianmrodri commented:
I added the required changes for Slither. It fixes all the Reentrancy warnings by moving `_afterBuyTransfer` before the payment in each method. Also renamed it to `_transfer` which is more accurate now.

The trick is that to be able to run Slither I had to create a non-upgradable version adding the constructors back. I named it `ERC721SimplePlacementsV1Slither.sol` not sure if it should be added and maintained in the repo as well.

ilanolkies commented on addresses.json:
```diff
@@ -0,0 +1,14 @@
+{
+    "ProxyAdmin": {
+      "rskMainnet": "0x0000000000000000000000000000000000000000",
+      "rskTestnet": "0x0000000000000000000000000000000000000000"
```
```suggestion
      "rskMainnet": "INVALID",
      "rskTestnet": "INVALID"
```

Using a valid zero address is error-prone

julianmrodri replied:
Good catch. Totally agree.

ilanolkies commented on addresses.json:
```diff
@@ -0,0 +1,14 @@
+      "rskTestnet": "0xbb71c17b28baf2b9e7f5b31660ecf758113b3fef"
+    },
+    "ERC721SimplePlacementsV1": {
+      "rskMainnet": "0x0000000000000000000000000000000000000000",
+      "rskTestnet": "0x0000000000000000000000000000000000000000"
```
```suggestion
      "rskMainnet": "INVALID",
      "rskTestnet": "INVALID"
```

ilanolkies commented on contracts/Migrations.sol:
```diff
@@ -1,8 +1,8 @@
 contract Migrations {
   address public owner;
-  uint public last_completed_migration;
+  uint public lastCompletedMigration;
```
Maybe we can ignore this file at all. It is not necessary to check it.

ilanolkies commented on contracts/ERC721SimplePlacementsV1.sol:
```diff
@@ -120,46 +126,53 @@ contract ERC721SimplePlacements is Context, ERC677TransferReceiver, IERC777Recip
 
     // With ERC-677
-    function tokenFallback(address from, uint256 /* amount */, bytes calldata data) external returns (bool) {
+    function tokenFallback(address from, uint256 /* amount */, bytes calldata data) external whenNotPaused returns (bool) {
```
Maybe the `whenNotPaused` works fine in `_transfer` method. Tiny refactor.

julianmrodri replied:
As discussed I will leave i at th etop level which is more clear in terms of understanding.

ilanolkies commented on migrations/2_deploy_nfts_contract.js:
```diff
@@ -0,0 +1,64 @@
+      console.log('Resulting querying implementation address: ' + implementationAddress);
+    });
+  }
+};
```
```suggestion
};

```

ilanolkies commented on slither.config.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "filter_paths": "node_modules"
+  }
```
```suggestion
}

```

julianmrodri commented:
Thanks for the review good points. Implemented all of them as discussed.
